### PR TITLE
ddl: support drop indexes for multi-schema change

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/14766

Problem Summary:

This PR supports drop indexes for multi-schema change.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: ddl/index.go [hunk 1]
```diff
@@ -721,38 +713,77 @@ func onDropIndex(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, _ error) {
...
+func removeDependentHiddenColumns(tblInfo *model.TableInfo, idxInfo *model.IndexInfo) {
+	hiddenColOffs := make([]int, 0)
+	for _, indexColumn := range idxInfo.Columns {
+		col := tblInfo.Columns[indexColumn.Offset]
+		if col.Hidden {
+			hiddenColOffs = append(hiddenColOffs, col.Offset)
+		}
+	}
+	// Sort the offset in descending order.
+	sort.Slice(hiddenColOffs, func(i, j int) bool {
```
Comment (hawkingrei) [inline ddl/index.go:725]: ```slices.SortFunc``` is better than ```sort.Slice```

1、it use pqqsort. it has better performance that quicksort of ````sort.Slice````
2、LessFunc avoid Bounds Check.

File: ddl/ddl_api.go [hunk 2]
```diff
@@ -3078,7 +3078,7 @@ func checkMultiSpecs(sctx sessionctx.Context, specs []*ast.AlterTableSpec) error
 func allSupported(specs []*ast.AlterTableSpec) bool {
 	for _, s := range specs {
 		switch s.Tp {
-		case ast.AlterTableAddColumns, ast.AlterTableDropColumn:
+		case ast.AlterTableAddColumns, ast.AlterTableDropColumn, ast.AlterTableDropIndex:
```
Comment (wjhuang2016) [inline ddl/ddl_api.go:3081]: AlterTableDropPrimaryKey？

Comment (tangenta) [reply ddl/ddl_api.go]: Updated, thanks.

**Review by zimulala — APPROVED** [review]
LGTM

**Review by wjhuang2016 — APPROVED** [review]

File: ddl/rollingback.go [hunk 3]
```diff
@@ -60,8 +60,8 @@ func convertAddIdxJob2RollbackJob(d *ddlCtx, t *meta.Meta, job *model.Job, tblIn
 		}
 	}
 
-	// the second args will be used in onDropIndex.
-	job.Args = []interface{}{indexInfo.Name, getPartitionIDs(tblInfo)}
+	// the second and the third args will be used in onDropIndex.
+	job.Args = []interface{}{indexInfo.Name, false /* ifExists */, getPartitionIDs(tblInfo)}
```
Comment (xhebox) [inline ddl/rollingback.go:64]: Won't this break the compatibility, if it is inserted into the middle, @tangenta ?

Comment (tangenta) [reply ddl/rollingback.go:64]: For `DropColumn()`, the `exists` argument is the last one... there are two cases:

- New version TiDB creates a job (args: [idx_name, ifExists]), old version TiDB runs the job(ifExists is ignored). 
- Old version TiDB creates a job(args: [idx_name]), new version TiDB runs the job(ifExists is decoded to `false`).

In the worse case, the DDL owner TiDB reports an error. The data is consistent.

**Pull request merged**